core/src/main/scala/kanaloa/reactive/dispatcher/queue/Worker.scala

```diff
@@ -7,186 +7,170 @@ import kanaloa.reactive.dispatcher.ApiProtocol.{QueryStatus, WorkFailed, WorkTim
+    case Terminated(`queue`)  ⇒ finish()
 
   }
 
-  def finish(): Unit = context stop self
-
-  def working(outstanding: Outstanding): Receive = ({
-    case Hold(period)        ⇒ delayBeforeNextWork = Some(period)
+  //in this state we are we waiting for 2 things to happen, Unregistration and Work completing
+  //the Worker will shift its state based on which one happens first
+  def unregisteringOutstanding(outstanding: Outstanding): Receive = routeeResponse(outstanding, becomeUnregistering) orElse {
+    case qs: QueryStatus     ⇒ qs reply UnregisteringOutstanding
 
-    case Terminated(`queue`) ⇒ context become retiring(Some(outstanding))
+    //ignore these
+    case Hold(x)             ⇒
+    case Retire              ⇒
+    case NoWorkLeft          ⇒
 
-    case Terminated(r) if r == routee ⇒
-      outstanding.fail(WorkFailed(s"due ${routee.path} is terminated"))
-      context become waitingForRoutee(None)
-      retrieveRoutee()
-
-    case qs: QueryStatus ⇒ qs reply Working
-
-    case Worker.Retire   ⇒ context become retiring(Some(outstanding))
-
-  }: Receive).orElse(waitingResult(outstanding, false))
-
-  def retiring(outstanding: Option[Outstanding]): Receive = ({
-    case Terminated(_)   ⇒ //ignore when retiring
-    case qs: QueryStatus ⇒ qs reply Retiring
-    //TODO: >IF< outstanding was populated, this would drop work, however that doesn't appear to be a valid state, maybe refactor the retiring state a bit
-    case Unregistered    ⇒ finish()
-    case Retire          ⇒ //already retiring
-    case Hold(period)    ⇒ //ignore
-
-  }: Receive) orElse (
-    if (outstanding.isDefined)
-      waitingResult(outstanding.get, true)
-    else {
-      case w: Work ⇒
-        sender ! Rejected(w, "Retiring")
-        finish()
-    }
-  )
+    case w: Work             ⇒ sender ! Rejected(w, "Retiring") //safety first
 
-  def waitingResult(
-    outstanding: Outstanding,
-    isRetiring:  Boolean
-  ): Receive = {
-    val handleResult: Receive =
-      (resultChecker orElse ({
+    //Either Unregistration completed, or the Queue died, in either way, we just need to wait for Work to finish
+    case Unregistered        ⇒ context become waitingToTerminate(outstanding)
+    case Terminated(`queue`) ⇒ context become waitingToTerminate(outstanding)
 
-        case m ⇒ Left(s"Unmatched Result '${descriptionOf(m)}' from the backend service, update your ResultChecker if you want to prevent it from being treated as an error.")
+    //work completed on way or another, just waiting for the Unregister ack
+    case WorkFinished        ⇒ context become unregistering
+  }
 
-      }: ResultChecker)).andThen[Unit] {
+  def becomeUnregistering(): Unit = {
+    queue ! Unregister(self)
+    context become unregistering
+  }
 
-        case Right(result) ⇒
-          outstanding.success(result)
-          workFinished(isRetiring)
+  //onRouteeFailure is what gets called if while waiting for a Routee response, the Routee dies.
```

Comment:
I had to pass in this `onRouteeFailure` function as when I first tried this all out, this code was capturing all of the `Terminated(routee)` stuff, and was handling as it if it was the response from the Routee.  In order to fix that, and _try_ to keep it straightforward, I have this function handling that event, and then calling the functin that each state passes in.  Currently, these are only state transitions.